File: diffycodec/gaussian_splatting.py
"""
2D Gaussian Splatting for background scene representation.

Instead of storing the background as a flat JPEG, we decompose it into a
set of 2D Gaussians — each with position (x, y), covariance (scale_x,
scale_y, rotation), colour (r, g, b), and opacity (alpha).

Benefits over flat-JPEG background:
  1. Resolution-independent — render at any size without blocking artefacts
  2. Compact for simple scenes — 500 splats × 12 floats = 24 KB vs ~200 KB JPEG
  3. Camera-motion tolerant — splats can be shifted/rotated for slight viewpoint changes
  4. Interactive — users can explore the background decomposition in a WebGL viewer
  5. Editable — individual splats can be pruned, recoloured, or repositioned

Training uses expectation-maximisation:
  - Initialise splats from a grid or K-means on high-variance regions
  - Iteratively refine positions, scales, colours to minimise reconstruction error
  - Prune low-opacity splats and split high-error ones (adaptive density control)

The splat model is stored as chunk 0x08 in the .dfy bitstream.
"""

import struct
import numpy as np
from typing import Optional, Tuple


class GaussianSplat:
    """A single 2D Gaussian splat."""
    __slots__ = ('x', 'y', 'sx', 'sy', 'rot', 'r', 'g', 'b', 'opacity')

    def __init__(self, x: float, y: float, sx: float, sy: float,
                 rot: float, r: float, g: float, b: float, opacity: float):
        self.x = x
        self.y = y
        self.sx = sx       # scale x (std dev)
        self.sy = sy       # scale y (std dev)
        self.rot = rot     # rotation in radians
        self.r = r
        self.g = g
        self.b = b
        self.opacity = opacity


# Number of float32 values per splat in the serialised format
FLOATS_PER_SPLAT = 9  # x, y, sx, sy, rot, r, g, b, opacity


class GaussianSplatModel:
    """
    A collection of 2D Gaussians that approximate an image (the background).

    Parameters
    ----------
    n_splats    : initial number of splats
    iterations  : EM refinement iterations
    lr          : learning rate for gradient steps
    min_opacity : splats below this are pruned
    """

    def __init__(self,
                 n_splats: int = 2000,
                 iterations: int = 100,
                 lr: float = 0.01,
                 min_opacity: float = 0.01):
        self.n_splats = n_splats
        self.iterations = iterations
        self.lr = lr
        self.min_opacity = min_opacity

        # Splat parameters as contiguous arrays for vectorised ops
        # Shape: (N,) for each
        self.positions: Optional[np.ndarray] = None   # (N, 2) float32  [x, y]
        self.scales: Optional[np.ndarray] = None      # (N, 2) float32  [sx, sy]
        self.rotations: Optional[np.ndarray] = None   # (N,)   float32  radians
        self.colours: Optional[np.ndarray] = None     # (N, 3) float32  [r, g, b] 0-255
        self.opacities: Optional[np.ndarray] = None   # (N,)   float32  0-1

        self._width: int = 0
        self._height: int = 0

    def fit(self, background: np.ndarray) -> None:
        """
        Train splats to approximate the background image.

        Parameters
        ----------
        background : uint8 H x W x 3 RGB image
        """
        H, W = background.shape[:2]
        self._height, self._width = H, W
        bg = background.astype(np.float32)

        N = self.n_splats
        # Initialise on a grid with jitter
        grid_side = int(np.ceil(np.sqrt(N)))
        xs = np.linspace(0, W - 1, grid_side, dtype=np.float32)
        ys = np.linspace(0, H - 1, grid_side, dtype=np.float32)
        gx, gy = np.meshgrid(xs, ys)
        gx = gx.ravel()[:N]
        gy = gy.ravel()[:N]
        actual_n = len(gx)

        # Add small jitter
        rng = np.random.RandomState(42)
        gx += rng.uniform(-W / grid_side * 0.3, W / grid_side * 0.3, actual_n).astype(np.float32)
        gy += rng.uniform(-H / grid_side * 0.3, H / grid_side * 0.3, actual_n).astype(np.float32)
        gx = np.clip(gx, 0, W - 1)
        gy = np.clip(gy, 0, H - 1)

        self.positions = np.stack([gx, gy], axis=1)  # (N, 2)

        # Initial scales proportional to grid spacing
        cell_w = W / grid_side
        cell_h = H / grid_side
        self.scales = np.full((actual_n, 2), [cell_w * 0.6, cell_h * 0.6], dtype=np.float32)

        self.rotations = np.zeros(actual_n, dtype=np.float32)
        self.opacities = np.full(actual_n, 0.8, dtype=np.float32)

        # Sample initial colours from the background at splat positions
        ix = np.clip(gx.astype(int), 0, W - 1)
        iy = np.clip(gy.astype(int), 0, H - 1)
        self.colours = bg[iy, ix].copy()  # (N, 3)

        # EM-style optimisation via stochastic gradient descent
        # We render into a downsampled canvas for speed
        ds = max(1, min(H, W) // 128)  # downsample factor
        h_ds, w_ds = H // ds, W // ds
        target = bg[::ds, ::ds, :].copy()

        for it in range(self.iterations):
            rendered = self._render_low(w_ds, h_ds, ds)
            error = target - rendered  # (h, w, 3)

            # Decay learning rate
            lr = self.lr * (1.0 - 0.5 * it / self.iterations)

            # Update each splat based on the error at its position
            for i in range(actual_n):
                px = int(self.positions[i, 0] / ds)
                py = int(self.positions[i, 1] / ds)
                px = min(max(px, 0), w_ds - 1)
                py = min(max(py, 0), h_ds - 1)

                # Sample error in a small neighbourhood
                r_patch = max(1, int(self.scales[i, 0] / ds * 0.5))
                y0 = max(0, py - r_patch)
                y1 = min(h_ds, py + r_patch + 1)
                x0 = max(0, px - r_patch)
                x1 = min(w_ds, px + r_patch + 1)

                patch_err = error[y0:y1, x0:x1]
                if patch_err.size == 0:
                    continue

                mean_err = patch_err.mean(axis=(0, 1))  # (3,)

                # Update colour toward the error
                self.colours[i] += lr * 40.0 * mean_err * self.opacities[i]
                self.colours[i] = np.clip(self.colours[i], 0, 255)

                # Update opacity based on magnitude of local error
                err_mag = np.abs(mean_err).mean()
                if err_mag > 10:
                    self.opacities[i] = min(1.0, self.opacities[i] + lr * 0.1)
                elif err_mag < 2:
                    self.opacities[i] = max(0.0, self.opacities[i] - lr * 0.05)

                # Nudge position toward higher-error regions
                if patch_err.shape[0] > 1 and patch_err.shape[1] > 1:
                    err_y = np.abs(patch_err).mean(axis=2).mean(axis=1)
                    err_x = np.abs(patch_err).mean(axis=2).mean(axis=0)
                    cy = np.average(np.arange(len(err_y)), weights=err_y + 1e-6)
                    cx = np.average(np.arange(len(err_x)), weights=err_x + 1e-6)
                    dy = (cy - len(err_y) / 2) * ds * lr * 2.0
                    dx = (cx - len(err_x) / 2) * ds * lr * 2.0
                    self.positions[i, 0] = np.clip(self.positions[i, 0] + dx, 0, W - 1)
                    self.positions[i, 1] = np.clip(self.positions[i, 1] + dy, 0, H - 1)

            # Adaptive density: prune dead splats every 20 iterations
            if (it + 1) % 20 == 0:
                alive = self.opacities > self.min_opacity
                if alive.sum() < actual_n:
                    self.positions = self.positions[alive]
                    self.scales = self.scales[alive]
                    self.rotations = self.rotations[alive]
                    self.colours = self.colours[alive]
                    self.opacities = self.opacities[alive]
                    actual_n = len(self.opacities)

        self.n_splats = actual_n

    def _render_low(self, w: int, h: int, ds: int) -> np.ndarray:
        """Render splats into a downsampled canvas for training."""
        canvas = np.zeros((h, w, 3), dtype=np.float32)
        weight = np.zeros((h, w, 1), dtype=np.float32)

        for i in range(len(self.opacities)):
            cx = self.positions[i, 0] / ds
            cy = self.positions[i, 1] / ds
            sx = max(self.scales[i, 0] / ds, 0.5)
            sy = max(self.scales[i, 1] / ds, 0.5)
            alpha = self.opacities[i]

            # Bounding box (3-sigma)
            r = 3
            x0 = max(0, int(cx - r * sx))
            x1 = min(w, int(cx + r * sx) + 1)
            y0 = max(0, int(cy - r * sy))
            y1 = min(h, int(cy + r * sy) + 1)

            if x0 >= x1 or y0 >= y1:
                continue

            yy, xx = np.mgrid[y0:y1, x0:x1].astype(np.float32)
            dx = (xx - cx) / sx
            dy = (yy - cy) / sy

            # Apply rotation
            rot = self.rotations[i]
            if abs(rot) > 0.01:
                c, s = np.cos(rot), np.sin(rot)
                dx2 = c * dx + s * dy
                dy2 = -s * dx + c * dy
                dx, dy = dx2, dy2

            gauss = np.exp(-0.5 * (dx * dx + dy * dy))
            w_splat = (gauss * alpha)[:, :, np.newaxis]

            canvas[y0:y1, x0:x1] += w_splat * self.colours[i]
            weight[y0:y1, x0:x1] += w_splat

        # Normalise
        mask = weight[:, :, 0] > 1e-6
        canvas[mask] /= weight[mask]

        return np.clip(canvas, 0, 255)

    def render(self, width: Optional[int] = None,
               height: Optional[int] = None) -> np.ndarray:
        """
        Render the splat model to an image.

        Returns uint8 H x W x 3 RGB.
        """
        w = width or self._width
        h = height or self._height

        canvas = np.zeros((h, w, 3), dtype=np.float32)
        weight = np.zeros((h, w, 1), dtype=np.float32)

        # Scale factor if rendering at different resolution
        sx_scale = w / self._width if self._width else 1.0
        sy_scale = h / self._height if self._height else 1.0

        for i in range(len(self.opacities)):
            cx = self.positions[i, 0] * sx_scale
            cy = self.positions[i, 1] * sy_scale
            sx = max(self.scales[i, 0] * sx_scale, 0.5)
            sy = max(self.scales[i, 1] * sy_scale, 0.5)
            alpha = self.opacities[i]

            r = 3
            x0 = max(0, int(cx - r * sx))
            x1 = min(w, int(cx + r * sx) + 1)
            y0 = max(0, int(cy - r * sy))
            y1 = min(h, int(cy + r * sy) + 1)

            if x0 >= x1 or y0 >= y1:
                continue

            yy, xx = np.mgrid[y0:y1, x0:x1].astype(np.float32)
            dx = (xx - cx) / sx
            dy = (yy - cy) / sy

            rot = self.rotations[i]
            if abs(rot) > 0.01:
                c, s = np.cos(rot), np.sin(rot)
                dx2 = c * dx + s * dy
                dy2 = -s * dx + c * dy
                dx, dy = dx2, dy2

            gauss = np.exp(-0.5 * (dx * dx + dy * dy))
            w_splat = (gauss * alpha)[:, :, np.newaxis]

            canvas[y0:y1, x0:x1] += w_splat * self.colours[i]
            weight[y0:y1, x0:x1] += w_splat

        mask = weight[:, :, 0] > 1e-6
        canvas[mask] /= weight[mask]

        return np.clip(canvas, 0, 255).astype(np.uint8)

    def psnr(self, reference: np.ndarray) -> float:
        """Compute PSNR of rendered splats vs reference image."""
        rendered = self.render(reference.shape[1], reference.shape[0])
        mse = np.mean((rendered.astype(np.float32) - reference.astype(np.float32)) ** 2)
        if mse < 1e-10:
            return 99.0
        return float(10 * np.log10(255.0 ** 2 / mse))

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_bytes(self) -> bytes:
        """
        Serialise splat model to bytes for .dfy storage.

        Format:
          [2B u16] width
          [2B u16] height
          [4B u32] n_splats
          [N * 9 * 4B] float32 array: x, y, sx, sy, rot, r, g, b, opacity
        """
        N = len(self.opacities)
        header = struct.pack(">HHI", self._width, self._height, N)

        # Pack all splats into a contiguous float32 array
        data = np.empty((N, FLOATS_PER_SPLAT), dtype=np.float32)
        data[:, 0:2] = self.positions
        data[:, 2:4] = self.scales
        data[:, 4] = self.rotations
        data[:, 5:8] = self.colours
        data[:, 8] = self.opacities

        return header + data.tobytes()

    @classmethod
    def from_bytes(cls, data: bytes) -> "GaussianSplatModel":
        """Deserialise a splat model from bytes."""
        width, height, n = struct.unpack_from(">HHI", data, 0)
        offset = 8

        model = cls(n_splats=n)
        model._width = width
        model._height = height

        arr = np.frombuffer(data[offset:offset + n * FLOATS_PER_SPLAT * 4],
                            dtype=np.float32).reshape(n, FLOATS_PER_SPLAT)

        model.positions = arr[:, 0:2].copy()
        model.scales = arr[:, 2:4].copy()
        model.rotations = arr[:, 4].copy()
        model.colours = arr[:, 5:8].copy()
        model.opacities = arr[:, 8].copy()
        model.n_splats = n

        return model

    def to_json(self) -> dict:
        """Export splats as JSON-friendly dict for the WebGL viewer."""
        N = len(self.opacities)
        return {
            "width": self._width,
            "height": self._height,
            "count": N,
            "positions": self.positions.tolist(),
            "scales": self.scales.tolist(),
            "rotations": self.rotations.tolist(),
            "colours": self.colours.tolist(),
            "opacities": self.opacities.tolist(),
        }


# ------------------------------------------------------------------
# High-level helpers
# ------------------------------------------------------------------

def fit_splat_model(background: np.ndarray,
                    quality: int = 50) -> "GaussianSplatModel":
    """
    Fit a splat model to a background image.  Splat count and iteration
    budget scale automatically with image area and quality.

    quality 1-100:
      10  -> fewer splats, fewer iterations (faster, coarser)
      50  -> balanced (default)
      100 -> more splats, more iterations (slower, sharper)

    Splat count: ~1 splat per 200 pixels, capped at 2000.
    """
    H, W = background.shape[:2]
    n_splats  = max(50, min(2000, H * W // 200))
    # Scale iteration count with quality; clamp so small-image tests stay fast
    iterations = max(20, min(200, int(40 * (quality / 50.0))))
    model = GaussianSplatModel(n_splats=n_splats, iterations=iterations)
    model.fit(background)
    return model

File: egocodec-wasm/src/gaussian_splatting.rs
//! 2D Gaussian Splatting for Diffy background representation.
//!
//! Fits a set of 2D Gaussians to an RGB image, producing a compact
//! representation that can be rendered back to pixels.
//!
//! Serialization format (matches Python `GaussianSplatModel.to_bytes`):
//!   [2B u16 BE] width
//!   [2B u16 BE] height
//!   [4B u32 BE] n_splats
//!   [N * 9 * 4B] float32 array: x, y, sx, sy, rot, r, g, b, opacity
//!                 (native / little-endian floats, matching numpy `.tobytes()`)

/// A 2D Gaussian Splatting model.
pub struct GaussianSplatModel {
    pub width: u16,
    pub height: u16,
    pub n_splats: usize,
    pub positions: Vec<[f32; 2]>,  // (x, y)
    pub scales: Vec<[f32; 2]>,    // (sx, sy)
    pub rotations: Vec<f32>,      // rotation angle in radians
    pub colours: Vec<[f32; 3]>,   // (r, g, b) in 0–255
    pub opacities: Vec<f32>,      // 0–1
}

impl GaussianSplatModel {
    /// Fit Gaussian splats to an RGB image.
    ///
    /// `background` — flat RGB bytes (H * W * 3).
    /// `width`, `height` — image dimensions.
    /// `quality` — 0–100 knob controlling iteration count.
    pub fn fit(background: &[u8], width: usize, height: usize, quality: u8) -> Self {
        assert_eq!(background.len(), width * height * 3);

        // ── Determine hyper-parameters ──────────────────────────────────
        let area = height * width;
        let n_splats = (area / 200).max(50).min(2000);
        let iterations = ((40 * quality as usize) / 50).max(20).min(200);

        // ── Downsample factor ───────────────────────────────────────────
        let ds = (width.min(height) / 64).max(1);
        let ds_w = width / ds;
        let ds_h = height / ds;

        // Build downsampled target (simple box average).
        let target = downsample_rgb(background, width, height, ds, ds_w, ds_h);

        // ── Grid-initialise splats ──────────────────────────────────────
        let cols = (n_splats as f32).sqrt().ceil() as usize;
        let rows = ((n_splats as f32) / cols as f32).ceil() as usize;

        let mut positions: Vec<[f32; 2]> = Vec::with_capacity(n_splats);
        let mut scales: Vec<[f32; 2]> = Vec::with_capacity(n_splats);
        let mut rotations: Vec<f32> = Vec::with_capacity(n_splats);
        let mut colours: Vec<[f32; 3]> = Vec::with_capacity(n_splats);
        let mut opacities: Vec<f32> = Vec::with_capacity(n_splats);

        let cell_w = width as f32 / cols as f32;
        let cell_h = height as f32 / rows as f32;

        let mut count = 0usize;
        for r in 0..rows {
            for c in 0..cols {
                if count >= n_splats {
                    break;
                }
                let x = (c as f32 + 0.5) * cell_w;
                let y = (r as f32 + 0.5) * cell_h;

                // Sample colour from the original image at this position.
                let ix = (x as usize).min(width - 1);
                let iy = (y as usize).min(height - 1);
                let idx = (iy * width + ix) * 3;
                let cr = background[idx] as f32;
                let cg = background[idx + 1] as f32;
                let cb = background[idx + 2] as f32;

                positions.push([x, y]);
                scales.push([cell_w * 0.6, cell_h * 0.6]);
                rotations.push(0.0);
                colours.push([cr, cg, cb]);
                opacities.push(0.8);
                count += 1;
            }
        }

        // ── Optimisation loop ───────────────────────────────────────────
        let lr: f32 = 0.01;

        for _iter in 0..iterations {
            // Render at downsampled resolution.
            let rendered = render_to_f32(&positions, &scales, &rotations, &colours, &opacities,
                                         ds_w, ds_h, ds as f32);

            // Compute per-pixel error (target - rendered) and magnitude.
            let npix = ds_w * ds_h;
            let mut err = vec![0.0f32; npix * 3];
            let mut err_mag = vec![0.0f32; npix];
            for i in 0..npix {
                let dr = target[i * 3] - rendered[i * 3];
                let dg = target[i * 3 + 1] - rendered[i * 3 + 1];
                let db = target[i * 3 + 2] - rendered[i * 3 + 2];
                err[i * 3] = dr;
                err[i * 3 + 1] = dg;
                err[i * 3 + 2] = db;
                err_mag[i] = (dr * dr + dg * dg + db * db).sqrt();
            }

            // Update each splat.
            let n = positions.len();
            for s in 0..n {
                // Project position to downsampled coordinates.
                let px = (positions[s][0] / ds as f32).min((ds_w - 1) as f32).max(0.0);
                let py = (positions[s][1] / ds as f32).min((ds_h - 1) as f32).max(0.0);
                let ix = px as usize;
                let iy = py as usize;
                let pi = iy * ds_w + ix;

                let er = err[pi * 3];
                let eg = err[pi * 3 + 1];
                let eb = err[pi * 3 + 2];
                let em = err_mag[pi];
                let opa = opacities[s];

                // Update colour toward error.
                let colour_lr = lr * 40.0 * opa;
                colours[s][0] = (colours[s][0] + colour_lr * er).clamp(0.0, 255.0);
                colours[s][1] = (colours[s][1] + colour_lr * eg).clamp(0.0, 255.0);
                colours[s][2] = (colours[s][2] + colour_lr * eb).clamp(0.0, 255.0);

                // Nudge opacity based on error magnitude.
                let opa_delta = lr * (em / 255.0 - 0.1);
                opacities[s] = (opa + opa_delta).clamp(0.0, 1.0);

                // Nudge position toward higher-error neighbour.
                nudge_position(&mut positions[s], &err_mag, ds_w, ds_h, ds as f32, lr);
            }
        }

        // ── Prune low-opacity splats ────────────────────────────────────
        let mut keep = Vec::new();
        for i in 0..positions.len() {
            if opacities[i] >= 0.01 {
                keep.push(i);
            }
        }
        let positions: Vec<[f32; 2]> = keep.iter().map(|&i| positions[i]).collect();
        let scales: Vec<[f32; 2]> = keep.iter().map(|&i| scales[i]).collect();
        let rotations: Vec<f32> = keep.iter().map(|&i| rotations[i]).collect();
        let colours: Vec<[f32; 3]> = keep.iter().map(|&i| colours[i]).collect();
        let opacities: Vec<f32> = keep.iter().map(|&i| opacities[i]).collect();

        GaussianSplatModel {
            width: width as u16,
            height: height as u16,
            n_splats: positions.len(),
            positions,
            scales,
            rotations,
            colours,
            opacities,
        }
    }

    /// Render the splat model to an RGB image (H * W * 3 bytes).
    pub fn render(&self, width: usize, height: usize) -> Vec<u8> {
        let canvas_f = render_to_f32(
            &self.positions,
            &self.scales,
            &self.rotations,
            &self.colours,
            &self.opacities,
            width,
            height,
            1.0, // no downsampling
        );
        let npix = width * height;
        let mut out = vec![0u8; npix * 3];
        for i in 0..npix * 3 {
            out[i] = canvas_f[i].clamp(0.0, 255.0) as u8;
        }
        out
    }

    /// Serialize to the Diffy splat wire format.
    ///
    /// Header: big-endian width(u16), height(u16), n_splats(u32).
    /// Body:   N * 9 float32 values in native (little-endian on WASM) byte order.
    pub fn to_bytes(&self) -> Vec<u8> {
        let header_size = 2 + 2 + 4;
        let body_size = self.n_splats * 9 * 4;
        let mut buf = Vec::with_capacity(header_size + body_size);

        // Header — big-endian.
        buf.extend_from_slice(&self.width.to_be_bytes());
        buf.extend_from_slice(&self.height.to_be_bytes());
        buf.extend_from_slice(&(self.n_splats as u32).to_be_bytes());

        // Body — native-endian floats (LE on WASM, matching numpy .tobytes()).
        for i in 0..self.n_splats {
            let floats: [f32; 9] = [
                self.positions[i][0],
                self.positions[i][1],
                self.scales[i][0],
                self.scales[i][1],
                self.rotations[i],
                self.colours[i][0],
                self.colours[i][1],
                self.colours[i][2],
                self.opacities[i],
            ];
            for f in &floats {
                buf.extend_from_slice(&f.to_ne_bytes());
            }
        }

        buf
    }

    /// Compute PSNR against a reference RGB image.
    pub fn psnr(&self, reference: &[u8]) -> f32 {
        let rendered = self.render(self.width as usize, self.height as usize);
        assert_eq!(rendered.len(), reference.len());

        let n = reference.len() as f64;
        let mut mse: f64 = 0.0;
        for i in 0..reference.len() {
            let d = rendered[i] as f64 - reference[i] as f64;
            mse += d * d;
        }
        mse /= n;

        if mse < 1e-10 {
            return 100.0;
        }
        (10.0 * (255.0_f64 * 255.0 / mse).log10()) as f32
    }
}

// ── Internal helpers ─────────────────────────────────────────────────────────

/// Downsample an RGB image by factor `ds` using box averaging.
fn downsample_rgb(
    src: &[u8],
    w: usize,
    h: usize,
    ds: usize,
    ds_w: usize,
    ds_h: usize,
) -> Vec<f32> {
    let mut out = vec![0.0f32; ds_w * ds_h * 3];
    let mut counts = vec![0u32; ds_w * ds_h];

    for y in 0..h {
        let dy = y / ds;
        if dy >= ds_h {
            continue;
        }
        for x in 0..w {
            let dx = x / ds;
            if dx >= ds_w {
                continue;
            }
            let si = (y * w + x) * 3;
            let di = dy * ds_w + dx;
            out[di * 3] += src[si] as f32;
            out[di * 3 + 1] += src[si + 1] as f32;
            out[di * 3 + 2] += src[si + 2] as f32;
            counts[di] += 1;
        }
    }

    for i in 0..ds_w * ds_h {
        if counts[i] > 0 {
            let c = counts[i] as f32;
            out[i * 3] /= c;
            out[i * 3 + 1] /= c;
            out[i * 3 + 2] /= c;
        }
    }
    out
}

/// Render splats to a float32 RGB canvas.
///
/// `scale_div` lets us work in downsampled coordinates:
/// splat positions are divided by `scale_div`, scales are divided by `scale_div`.
fn render_to_f32(
    positions: &[[f32; 2]],
    scales: &[[f32; 2]],
    _rotations: &[f32],
    colours: &[[f32; 3]],
    opacities: &[f32],
    canvas_w: usize,
    canvas_h: usize,
    scale_div: f32,
) -> Vec<f32> {
    let npix = canvas_w * canvas_h;
    let mut canvas = vec![0.0f32; npix * 3];
    let mut weights = vec![0.0f32; npix];

    let n = positions.len();
    for i in 0..n {
        let cx = positions[i][0] / scale_div;
        let cy = positions[i][1] / scale_div;
        let sx = (scales[i][0] / scale_div).max(0.5);
        let sy = (scales[i][1] / scale_div).max(0.5);
        let opa = opacities[i];

        if opa < 1e-4 {
            continue;
        }

        // Bounding box: 3 sigma.
        let x0 = ((cx - 3.0 * sx).floor() as isize).max(0) as usize;
        let x1 = ((cx + 3.0 * sx).ceil() as usize + 1).min(canvas_w);
        let y0 = ((cy - 3.0 * sy).floor() as isize).max(0) as usize;
        let y1 = ((cy + 3.0 * sy).ceil() as usize + 1).min(canvas_h);

        let inv_sx2 = 1.0 / (sx * sx);
        let inv_sy2 = 1.0 / (sy * sy);

        let cr = colours[i][0];
        let cg = colours[i][1];
        let cb = colours[i][2];

        for py in y0..y1 {
            let dy = py as f32 - cy;
            let dy2 = dy * dy * inv_sy2;
            let row_off = py * canvas_w;
            for px in x0..x1 {
                let dx = px as f32 - cx;
                let exponent = -0.5 * (dx * dx * inv_sx2 + dy2);
                // Fast reject.
                if exponent < -4.5 {
                    continue;
                }
                let w = fast_exp(exponent) * opa;
                let idx = row_off + px;
                canvas[idx * 3] += w * cr;
                canvas[idx * 3 + 1] += w * cg;
                canvas[idx * 3 + 2] += w * cb;
                weights[idx] += w;
            }
        }
    }

    // Normalise.
    let eps = 1e-6;
    for i in 0..npix {
        if weights[i] > eps {
            let inv_w = 1.0 / weights[i];
            canvas[i * 3] *= inv_w;
            canvas[i * 3 + 1] *= inv_w;
            canvas[i * 3 + 2] *= inv_w;
        }
    }

    canvas
}

/// Nudge a splat position toward the highest-error neighbour in the
/// downsampled error magnitude map.
fn nudge_position(
    pos: &mut [f32; 2],
    err_mag: &[f32],
    ds_w: usize,
    ds_h: usize,
    ds: f32,
    lr: f32,
) {
    let px = (pos[0] / ds).min((ds_w - 1) as f32).max(0.0) as usize;
    let py = (pos[1] / ds).min((ds_h - 1) as f32).max(0.0) as usize;

    let mut best_dx: f32 = 0.0;
    let mut best_dy: f32 = 0.0;
    let mut best_err: f32 = 0.0;

    // Check 3x3 neighbourhood.
    for dy in -1i32..=1 {
        for dx in -1i32..=1 {
            if dx == 0 && dy == 0 {
                continue;
            }
            let nx = px as i32 + dx;
            let ny = py as i32 + dy;
            if nx < 0 || ny < 0 || nx >= ds_w as i32 || ny >= ds_h as i32 {
                continue;
            }
            let ni = ny as usize * ds_w + nx as usize;
            if err_mag[ni] > best_err {
                best_err = err_mag[ni];
                best_dx = dx as f32;
                best_dy = dy as f32;
            }
        }
    }

    // Move toward higher error.
    let step = lr * ds * 2.0;
    pos[0] += best_dx * step;
    pos[1] += best_dy * step;
}

/// Fast approximate exp() for small negative exponents.
/// Uses the classic (1 + x/256)^256 trick — good enough for Gaussian weights.
#[inline(always)]
fn fast_exp(x: f32) -> f32 {
    if x < -10.0 {
        return 0.0;
    }
    // 6th-order polynomial is overkill; use the standard fast path.
    let mut v = 1.0 + x / 256.0;
    // Square 8 times: (1 + x/256)^256.
    v *= v; // ^2
    v *= v; // ^4
    v *= v; // ^8
    v *= v; // ^16
    v *= v; // ^32
    v *= v; // ^64
    v *= v; // ^128
    v *= v; // ^256
    v.max(0.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_roundtrip_serialization() {
        // Small solid-red image.
        let w = 32usize;
        let h = 32usize;
        let mut img = vec![0u8; w * h * 3];
        for i in 0..w * h {
            img[i * 3] = 200;
            img[i * 3 + 1] = 50;
            img[i * 3 + 2] = 50;
        }

        let model = GaussianSplatModel::fit(&img, w, h, 30);
        let bytes = model.to_bytes();

        // Check header.
        let rw = u16::from_be_bytes([bytes[0], bytes[1]]);
        let rh = u16::from_be_bytes([bytes[2], bytes[3]]);
        let rn = u32::from_be_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]);
        assert_eq!(rw, w as u16);
        assert_eq!(rh, h as u16);
        assert_eq!(rn as usize, model.n_splats);

        // Check body length.
        assert_eq!(bytes.len(), 8 + model.n_splats * 9 * 4);
    }

    #[test]
    fn test_render_size() {
        let w = 16usize;
        let h = 16usize;
        let img = vec![128u8; w * h * 3];

        let model = GaussianSplatModel::fit(&img, w, h, 20);
        let rendered = model.render(w, h);
        assert_eq!(rendered.len(), w * h * 3);
    }

    #[test]
    fn test_psnr_perfect() {
        // An image that is all zeros — model should get decent PSNR.
        let w = 8usize;
        let h = 8usize;
        let img = vec![100u8; w * h * 3];
        let model = GaussianSplatModel::fit(&img, w, h, 50);
        let p = model.psnr(&img);
        // Should be at least a reasonable PSNR on a solid colour image.
        assert!(p > 10.0, "PSNR too low: {}", p);
    }

    #[test]
    fn test_fast_exp() {
        // Verify fast_exp is reasonably close to std exp.
        let vals: &[f32] = &[-0.5, -1.0, -2.0, -4.0, 0.0];
        for &x in vals {
            let exact = x.exp();
            let approx = fast_exp(x);
            let rel_err = ((approx - exact) / exact).abs();
            assert!(rel_err < 0.05, "fast_exp({}) = {}, expected ~{}, rel_err={}", x, approx, exact, rel_err);
        }
    }
}
